Prototype regex literal AST and emission

hamishknight opened the pull request:
With `-enable-experimental-string-processing`, start lexing `'` delimiters as regex literals (this is just a placeholder delimiter for now). The contents of which gets passed to the libswift library, which can return an error string to be emitted, or null for success.

The libswift side isn't yet hooked up to the Swift regex parser, so for now just emit a dummy diagnostic for regexes starting with quantifiers.

If successful, build an AST node which will be emitted as an implicit call to an `init(_regexString:)` initializer of an in-scope `Regex` decl (which will eventually be a known stdlib decl).

hamishknight commented:
@swift-ci please test

milseman commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -91,6 +91,9 @@ ERROR(forbidden_interpolated_string,none,
 
+ERROR(regex_literal_parsing_error,none,
+      "%0", (StringRef))
+
```
This is ok for now, but will drop source locations on the floor. We'll want to figure out a way to communicate source ranges back.

beccadax replied:
We'll also want to be able to communicate back error codes, I think, so that diagnostics can be localized, customized, linked to educational notes, etc. And we'll want to be able to communicate multiple diagnostics, do fix-its, etc. But we should probably build all this up incrementally as we discover a need for it.

milseman commented on lib/Parse/Lexer.cpp:
```diff
@@ -1621,6 +1623,38 @@ static size_t commonPrefixLength(StringRef shorter, StringRef longer) {
+
+  formToken(tok::regex_literal, TokStart);
+}
+
```
(longer term) What do you think about pulling this functionality out of C++ and into the package? That is C++ passes in the character pointer and receives the terminating pointer for when it's done. That would allow us to mock up multiple kinds of literals, such as:

```
let regexen: Array<Regex> = [
  '/PCRE/',
  '|Modern|',
  '''
    multiline     // Line-ending comment
  ''',
  #'r'a'w'#
]
```  

```

hamishknight replied:
I think that would work, though it may be more complicated if we eventually decide to go with `/` delimiters, as that would likely require a bunch of lexing heuristics on the C++ side. But maybe that could be a special case, with other delimiters being passed straight through

milseman commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2110,6 +2113,56 @@ ParserResult<Expr> Parser::parseExprStringLiteral() {
+  // semantic expression for the regex. The type-checker will provide it with
+  // the correct contextual type. We delay the contextual type for a couple of
+  // reasons:
+  // 1. We need to delay type lookup until after parsing.
```
Eventually (hopefully soon) the parser will also communicate back info about the capture structures. Would this approach extend naturally to handle that?

hamishknight replied:
I think that would work fine, we could attach that info to the AST and resolve it when type-checking.

milseman commented on test/ExperimentalRegex/Parse/regex.swift:
```diff
@@ -0,0 +1,14 @@
+// RUN: %target-typecheck-verify-swift -enable-experimental-string-processing
```
CC @rxwei , who has been plumbing through the `_Regex` module.

rintaro commented on lib/Parse/Lexer.cpp:
```diff
@@ -30,17 +30,19 @@
-
-using namespace swift;
-using namespace swift::syntax;
+ParseRegexStrawperson Lexer::getExperimentalRegexParsingFn() {
+  return parseRegexStrawperson;
+}
```
At this point, this function and `Parser_registerParseRegexStrawperson` do nothing to do with `Lexer`. Can we move these to `Parser`. I guess we can implement this in `Parser.cpp` or maybe introduce `ParseRegex.cpp`.

hamishknight replied:
@rintaro I've moved the regex lexing and parsing logic into a ParseRegex.cpp file, that way we don't have to expose a method to grab the parsing function, does that look good to you?

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2110,6 +2113,56 @@ ParserResult<Expr> Parser::parseExprStringLiteral() {
+
+  auto loc = consumeToken();
+
+  // Create an implicit .init(_regexString: "<regex text>") call to serve as the
+  // semantic expression for the regex. The type-checker will provide it with
+  // the correct contextual type. We delay the contextual type for a couple of
```
Can we move this logic to `Sema` (e.g. `PreCheckExpression`)? In general, we want to move semantic logic out of `Parser` as much as possible.
I think `RegexExpr::createParsed` can initialize `SemanticExpr` with `nullptr`

hamishknight replied:
I'd rather not introduce mutability to the AST, but I guess this synthesis could be done lazily (maybe through a request)?

hamishknight replied:
@rxwei is planning on removing this in a follow-up (in favor of logic in CSApply + SILGen), so I'm going to leave it as-is for now

hamishknight commented:
https://github.com/apple/swift-syntax/pull/347

@swift-ci please test

milseman commented:
@swift-ci please test windows platform

hamishknight commented:
https://github.com/apple/swift-syntax/pull/347

@swift-ci please test

hamishknight committed:
Prototype regex literal AST and emission

With `-enable-experimental-string-processing`,
start lexing `'` delimiters as regex literals (this
is just a placeholder delimiter for now). The
contents of which gets passed to the libswift
library, which can return an error string to be
emitted, or null for success.

The libswift side isn't yet hooked up to the Swift
regex parser, so for now just emit a dummy
diagnostic for regexes starting with quantifiers.

If successful, build an AST node which will be
emitted as an implicit call to an
`init(_regexString:)` initializer of an in-scope
`Regex` decl (which will eventually be a known
stdlib decl).

hamishknight commented:
https://github.com/apple/swift-syntax/pull/347

@swift-ci please test
